nixos/modules/services/network-filesystems/samba.nix

```
@@ -3,232 +3,162 @@
-
   cfg = config.services.samba;
 
-  samba = cfg.package;
-
-  shareConfig = name:
-    let share = getAttr name cfg.shares; in
-    "[${name}]\n " + (smbToString (
-       map
-         (key: "${key} = ${smbToString (getAttr key share)}\n")
-         (attrNames share)
-    ));
-
-  configFile = pkgs.writeText "smb.conf"
-    (if cfg.configText != null then cfg.configText else
-    ''
-      [global]
-      security = ${cfg.securityType}
-      passwd program = /run/wrappers/bin/passwd %u
-      invalid users = ${smbToString cfg.invalidUsers}
-
-      ${cfg.extraConfig}
-
-      ${smbToString (map shareConfig (attrNames cfg.shares))}
-    '');
-
-  # This may include nss_ldap, needed for samba if it has to use ldap.
-  nssModulesPath = config.system.nssModules.path;
-
-  daemonService = appName: args:
-    { description = "Samba Service Daemon ${appName}";
-
-      after = [ (mkIf (cfg.enableNmbd && "${appName}" == "smbd") "samba-nmbd.service") "network.target" ];
-      requiredBy = [ "samba.target" ];
-      partOf = [ "samba.target" ];
-
-      environment = {
-        LD_LIBRARY_PATH = nssModulesPath;
-        LOCALE_ARCHIVE = "/run/current-system/sw/lib/locale/locale-archive";
-      };
-
-      serviceConfig = {
-        ExecStart = "${samba}/sbin/${appName} --foreground --no-process-group ${args}";
-        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
-        LimitNOFILE = 16384;
-        PIDFile = "/run/${appName}.pid";
-        Type = "notify";
-        NotifyAccess = "all"; #may not do anything...
-        Slice = "system-samba.slice";
-      };
-      unitConfig.RequiresMountsFor = "/var/lib/samba";
-
-      restartTriggers = [ configFile ];
-    };
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "smb.conf" cfg.settings;
 
 in
 
 {
+  meta = {
+    doc = ./samba.md;
+    maintainers = [ lib.maintainers.anthonyroussel ];
+  };
+
   imports = [
     (mkRemovedOptionModule [ "services" "samba" "defaultShare" ] "")
     (mkRemovedOptionModule [ "services" "samba" "syncPasswordsByPam" ] "This option has been removed by upstream, see https://bugzilla.samba.org/show_bug.cgi?id=10669#c10")
+
+    (lib.mkRemovedOptionModule [ "services" "samba" "configText" ] ''
+      Use services.samba.settings instead.
+
+      This is part of the general move to use structured settings instead of raw
+      text for config as introduced by RFC0042:
+      https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md
+    '')
+    (lib.mkRemovedOptionModule [ "services" "samba" "extraConfig" ] "Use services.samba.settings instead.")
+    (lib.mkRenamedOptionModule [ "services" "samba" "invalidUsers" ] [ "services" "samba" "settings" "global" "invalid users" ])
+    (lib.mkRenamedOptionModule [ "services" "samba" "securityType" ] [ "services" "samba" "settings" "global" "security type" ])
+    (lib.mkRemovedOptionModule [ "services" "samba" "shares" ] "Use services.samba.settings instead.")
```

```suggestion
    (lib.mkRenamedOptionModule [ "services" "samba" "shares" ] [ "services" "samba" "settings" ])
```

I think we can just rename shares to settings. As far as I see the format didn't change. This makes one less breaking change. This way only users that used `extraConfig` or `configText` are affected.

---

Thank you so much for the review @bachp! 

You're right! I changed that line to mkRenamedOptionModule.